[AutoDiff] Serialize derivative function configurations per module.

dan-zheng opened the pull request:
`@differentiable` and `@derivative` attributes register derivatives for
`AbstractFunctionDecl`s for a particular "derivative function configuration":
parameter indices and derivative generic signature.

To find `@derivative` functions registered in other Swift modules, derivative
function configurations must be serialized per module. When configurations for
a `AbstractFunctionDecl` are requested, all configurations from imported
modules are deserialized. This module serialization technique has precedent: it
is used for protocol conformances (e.g. extension declarations for a nominal
type) and Obj-C members for a class type.

Add `AbstractFunctionDecl::getDerivativeFunctionConfigurations` entry point
for accessing derivative function configurations.

Use `AbstractFunctionDecl::getDerivativeFunctionConfigurations` to
implement `findMinimalDerivativeConfiguration` for canonical derivative
function configuration lookup, replacing `getMinimalASTDifferentiableAttr`.

Unblocks TF-835: lowering `@derivative` attributes directly to SIL
differentiability witnesses without generating implicit `@differentiable`
attributes.

dan-zheng committed:
[AutoDiff] Serialize derivative function configurations per module.

`@differentiable` and `@derivative` attributes register derivatives for
`AbstractFunctionDecl`s for a particular "derivative function configuration":
parameter indices and dervative generic signature.

To find `@derivative` functions registered in other Swift modules, derivative
function configurations must be serialized per module. When configurations for
a `AbstractFunctionDecl` are requested, all configurations from imported
modules are deserialized. This module serialization technique has precedent: it
is used for protocol conformances (e.g. extension declarations for a nominal
type) and Obj-C members for a class type.

Add `AbstractFunctionDecl::getDerivativeFunctionConfigurations` entry point
for accessing derivative function configurations.

Use `AbstractFunctionDecl::getDerivativeFunctionConfigurations` to
implement `findMinimalDerivativeConfiguration` for canonical derivative
function configuration lookup, replacing `getMinimalASTDifferentiableAttr`.

Unblocks TF-815: lowering `@derivative` attributes directly to SIL
differentiability witnesses without generating implicit `@differentiable`
attributes.

dan-zheng commented on lib/AST/Decl.cpp:
```diff
@@ -6998,6 +6998,49 @@ StringRef AbstractFunctionDecl::getInlinableBodyText(
+  });
+}
+
+ArrayRef<AutoDiffConfig>
```
Note: even though `@differentiable` and `@derivative` attributes do not currently store result indices, I decided to use `AutoDiffConfig` (which contains result indices) to represent AST derivative function configurations. I also tried creating a new "ASTAutoDiffConfig" struct containing just parameter indices and derivative generic signature, but decided against the code duplication.

I think the direction is to eventually plumb result indices through the differentiation system (TF-1038), so preemptively using result indices (`AutoDiffConfig`) seems good.

marcrasi commented on include/swift/SILOptimizer/Utils/Differentiation/DerivativeLookup.h:
```diff
@@ -36,17 +36,26 @@ getExactDifferentiabilityWitness(SILModule &module, SILFunction *original,
+/// \param derivativeGenericSignature is an output parameter that is set to the
+/// derivative generic signature of the minimal configuration, or the `nullptr`
+/// if no such configuration exists.
+bool findMinimalDerivativeConfiguration(
```
This signature has become very complicated with lots of output parameters. Could it instead return a SIL-lowererd `Optional<AutoDiffConfig>`, plus an output parameter for `minimalASTParameterIndices`?

dan-zheng replied:
Nice idea! Done.

marcrasi commented on include/swift/AST/Decl.h:
```diff
@@ -5694,6 +5694,25 @@ class AbstractFunctionDecl : public GenericContext, public ValueDecl {
+  /// Prepare to traverse the list of derivative function configurations.
+  void prepareDerivativeFunctionConfigurations();
+
+  /// A uniqued list of derivative function configurations.
```
I think a note about who is responsible for populating this would be very helpful. It always takes me a while to figure out where these kinds of things get populated.

"Typechecking is responsible for populating derivative function configurations specified in the current module, and module loading is responsible for populating derivative function configurations specified in imported modules."

dan-zheng replied:
Good idea! Done in https://github.com/apple/swift/pull/28608/commits/11c4db418e2a217a2392461e4f471a02750cd82b.

marcrasi reviewed (approved):
Cooool!

dan-zheng committed:
Address review feedback.

dan-zheng commented on include/swift/AST/ModuleLoader.h:
```diff
@@ -151,6 +154,25 @@ class ModuleLoader {
+  /// generations.
+  virtual void loadDerivativeFunctionConfigurations(
+      AbstractFunctionDecl *originalAFD, unsigned previousGeneration,
+      llvm::SetVector<AutoDiffConfig> &results){};
```
I thought so too, but `clang-format` surprisingly removed the space. I added it back in https://github.com/apple/swift/pull/28608/commits/11c4db418e2a217a2392461e4f471a02750cd82b based on our personal taste and to match the surrounding code.

dan-zheng committed:
Address more review feedback.

Change `findMinimalDerivativeConfiguration` to return `Optional<AutoDiffConfig>`.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
Extended test suite passed. Merging to unblock progress.

rxwei commented:
Very nice. Thanks for pushing this through :)
